sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+        """
+        pass
+
+    def get_max_seq_len(self) -> Optional[int]:
+        """
+        :return: The maximum length supported by the block if such a restriction exists.
+        """
+        return None
+
+
+class StatelessBlockLayer(SharedEncoderDecoderLayer):
+    """
+    Runs the given block on the target side of decoder being applicable to any block which does not require to keep
+    any state during at inference time (when `decode_step` is called). The latter means that the block can be run as
+
+        concat(block(data[:,t,:]) for t in range(seq_len), dim=1)
+
+    Namely, at time t we only need the data point at time t from the previous layer.
+
+    """
+
+    def __init__(self, block: StatelessBlock):
+        self.block = block
+
+    def process_sequence(self,
+                         data: mx.sym.Symbol,
+                         lengths: mx.sym.Symbol,
+                         max_length: int) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        return self.block(data, lengths, max_length), lengths, max_length
+
+    def get_num_hidden(self) -> int:
+        return self.block.get_num_hidden()
+
+    def decode_step(self,
+                    step: int,
+                    source_encoded: mx.sym.Symbol,
+                    source_encoded_lengths: mx.sym.Symbol,
+                    source_encoded_max_length: int,
+                    target: mx.sym.Symbol,
+                    states: List[mx.sym.Symbol],
+                    att_dict: Dict[str, Dict[str, mx.sym.Symbol]]) -> Tuple[mx.sym.Symbol, List[mx.sym.Symbol]]:
+        # (batch_size, 1, num_hidden)
+        target = mx.sym.expand_dims(target, axis=1)
+
+        # (batch_size, 1, num_hidden_block)
+        hidden = self.block.step(step, target)
+
+        # (batch_size, num_hidden_block)
+        hidden = mx.sym.reshape(hidden, shape=(0, -1))
+        return hidden, []
+
+    def get_max_seq_len(self):
+        return self.block.get_max_seq_len()
+
+
+class FeedForwardBlock(StatelessBlock):
+    """
+    Position-wise feed-forward network with activation.
+    """
+
+    def __init__(self,
+                 num_hidden: int,
+                 dropout: float = 0.0,
+                 act_type: str = C.RELU,
+                 prefix: str = "") -> None:
+        self.num_hidden = num_hidden
+        self.dropout = dropout
+        self.prefix = prefix
+        self.act_type = act_type
+        self.w_i2h = mx.sym.Variable('%si2h_weight' % prefix)
+        self.b_i2h = mx.sym.Variable('%si2h_bias' % prefix)
+
+    def _pre_activation_num_hidden(self):
+        if self.act_type == C.GLU:
+            return 2 * self.num_hidden
+        else:
+            return self.num_hidden
+
+    def _ff(self, data: mx.sym.Symbol):
+        h = mx.sym.FullyConnected(data=data, num_hidden=self._pre_activation_num_hidden(),
+                                  weight=self.w_i2h, bias=self.b_i2h,
+                                  flatten=False, name=self.prefix + "_ff")
```

Comment:
our prefixes passed to classes typically already contain "_" right? (See line 523 for example)

Reply:
good point, removed.